### PR TITLE
Async File Writer (Buffer)

### PR DESCRIPTION
**Pull request by scottyob**
### Purpose

There have been a couple of times where the vario instruments go bad.  The theory is that something can back up the runloop and stop the more realtime sensors and instruments from running correctly.

We have also noticed that file.write calls can sometimes block for >900ms (not sure why, but we'll call this an *event*), and that the behavior of incorrect vario readings are more common when we're doing debug / bus logging.

This change introduces a new "asynchronous logger" library that will have a low priority task read from a ring buffer and write them to disk.  It also adds a 10-second task to write some statistics to our log file.

The statistics format is
```
S<timestamp>,queueLowWatermark,messagesDropped,fileWriteHighWatermarkUs
```

In the following tests, it should be noted that the max freesize is [always (buffer_size/2)-headersize ](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/system/freertos_additions.html#_CPPv425xRingbufferGetMaxItemSize15RingbufHandle_t)

### Test Plan

I started the debug log and left it running for a number of minutes.  The "log task" needs to have a stack size of about 4K, testing with 2K caused a buffer overflow / crash.

I also found that there were a number of these 913ms *events*.  With a 4K ring buffer, we managed to exhaust the buffer and still drop 40-50 of log messages (but turned this into a lossy log instead of causing back-pressure on the runloop)

```
S129999,2008,0,2139
S139999,2000,0,2130
S149999,0,45,911966
S159999,2008,45,3065
S169999,2008,45,2229
...
S489999,2012,45,2107
S499999,1984,45,2534
S509999,16,85,912034
S519999,2008,85,3069
...
S839999,1992,85,9479
S849999,2008,85,5593
S859999,2020,85,2121
S869999,0,120,913181
S879999,2008,120,3150
S889999,2008,120,2516
```

With an 8K buffer, the ~915ms file write event was able to absorb the logs and not drop any writes.
```
scott@Bethanys-MacBook-Air buslogs % grep "^S" BusLog_2080-01-05_2359.log
S9999,3972,0,3933
S19999,3952,0,3076
...
S279999,3956,0,5757
S289999,1852,0,914636
S299999,3952,0,2138
```

### Memory Usage
8KB for the circular buffer, 4KB for the log writer task causing a regression of about 12KB of RAM (free head with all bells and whistles running from 52KB to 40KB)

```
BEFORE:
=== Memory Stats ===
Total Heap: 272 KB
Free Heap: 52 KB
Used Heap: 219 KB
Largest Free Block: 30 KB
Minimum Free Heap Ever: 50 KB
Main Task Stack High Water Mark: 12 KB
Free PSRAM: 0 bytes
Largest Free PSRAM Block: 0 bytes
====================

AFTER:
=== Memory Stats ===
Total Heap: 272 KB
Free Heap: 40 KB
Used Heap: 232 KB
Largest Free Block: 30 KB
Minimum Free Heap Ever: 37 KB
Main Task Stack High Water Mark: 12 KB
Free PSRAM: 0 bytes
Largest Free PSRAM Block: 0 bytes
====================
```

### CODE REVIEW

**Comment by scottyob** [comment]
This is to address #223

File: src/vario/logging/async_logger.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+ private:
+  static void writerTask(void* arg);
+  static RingbufHandle_t rb;
+  static TaskHandle_t task;
+  static size_t freeLowWaterMark;  // Size in bytes of available space in the buffer, low watermark
+  static size_t bufferSize;        // Size of the ring buffer
+  static uint32_t droppedEntries;  // Number of log messages dropped (buffer was full)
+  static unsigned long
+      writeHighWatermarkUs;  // High watermark of how long it took to write to the SD card.
```
Comment (BenjaminPelletier) [inline src/vario/logging/async_logger.h:56]: Can we have trailing underscores in the names of private and protected fields to set them apart from local variables?

```suggestion
 private:
  static void writerTask(void* arg);
  static RingbufHandle_t rb_;
  static TaskHandle_t task_;
  static size_t freeLowWaterMark_;  // Size in bytes of available space in the buffer, low watermark
  static size_t bufferSize_;        // Size of the ring buffer
  static uint32_t droppedEntries_;  // Number of log messages dropped (buffer was full)
  static unsigned long
      writeHighWatermarkUs_;  // High watermark of how many microseconds a single log entry took to write to the SD card.
```

File: src/vario/logging/async_logger.cpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    LogMsgHeader* msg = (LogMsgHeader*)xRingbufferReceive(rb, &item_len, portMAX_DELAY);
+    if (msg) {
+      auto header = (LogMsgHeader*)msg;
+      char* payload = (char*)(header + 1);
+
+      // Write the payload (just after the header) to the SD card.
+      if (header->file) {
+        // Write the payload to the file and time how long the call took.
+        auto before = micros();
+        header->file->write((uint8_t*)payload, header->len);
```
[the remark this answers is not in the document]
Comment (scottyob) [reply src/vario/logging/async_logger.cpp]: Seems like a pretty big refactor though.  At the moment, I believe FAT file operations are protected by a mutex (see https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/kconfig-reference.html#config-fatfs-timeout-ms) so we should be pretty safe.

We also purge the buffers before writing footers and closing the files.  It *seems* pretty safe to me as is.  If you wanted to refactor, mind commandeering the diff, as that change would be a little bit more effort than I wanted to put in for this one.
